contracts/utils/Strings.sol

```diff
@@ -115,4 +120,85 @@ library Strings {
     function equal(string memory a, string memory b) internal pure returns (bool) {
         return bytes(a).length == bytes(b).length && keccak256(bytes(a)) == keccak256(bytes(b));
     }
+
+    /**
+     * @dev Parse an decimal string and returns the value as a `uint256`.
+     *
+     * This function will revert if:
+     * - the string contains any character that is not in [0-9].
+     * - the result does not fit in a uint256.
+     */
+    function toUint(string memory input) internal pure returns (uint256) {
+        bytes memory buffer = bytes(input);
+
+        uint256 result = 0;
+        for (uint256 i = 0; i < buffer.length; ++i) {
+            result *= 10; // will revert if overflow
+            result += _parseChr(buffer[i], 10);
+        }
+        return result;
+    }
```

Comment:
this would reduce gas by 3.36% (444 gas) on avg. for `toUint`
```suggestion
    function toUint(string memory input) internal pure returns (uint256 result) {
        bytes memory buffer = bytes(input);

        uint256 bufferLength = buffer.length;
        for (uint256 i = 0; i < bufferLength; ++i) {
            result *= 10; // will revert if overflow
            result += _parseChr(buffer[i], 10);
        }
    }
```

Reply:
Interresting. The compiler should optimize "length caching" by itself.

Reply:
i'm not aware it caches the length yet -- it does the optimization to `i++` unchecked as part of Solidity 0.8.22 tho. here's the gas diff on `toUint` before and after using the buffer length cached "manually":

before:
| test/utils/Strings.t.sol:Gas contract |                 |       |        |       |         |
|---------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                       | Deployment Size |       |        |       |         |
| 220588                                | 803             |       |        |       |         |
| Function Name                         | min             | avg   | median | max   | # calls |
| toUint                                | 1133            | 14165 | 10693  | 37951 | 256     |

after:
Suite result: ok. 4 passed; 0 failed; 0 skipped; finished in 514.46ms (1.41s CPU time)
| test/utils/Strings.t.sol:Gas contract |                 |       |        |       |         |
|---------------------------------------|-----------------|-------|--------|-------|---------|
| Deployment Cost                       | Deployment Size |       |        |       |         |
| 221452                                | 807             |       |        |       |         |
| Function Name                         | min             | avg   | median | max   | # calls |
| toUint                                | 1141            | 11482 | 4941   | 37728 | 256     |


function:
```solidity
    function toUint(string memory input) internal pure returns (uint256) {
        bytes memory buffer = bytes(input);
        uint256 bufferLength = buffer.length;

        uint256 result = 0;
        for (uint256 i = 0; i < bufferLength; ++i) {
            result *= 10; // will revert if overflow
            result += _parseChr(buffer[i], 10);
        }
        return result;
    }
```